### PR TITLE
Feature/5 3/introduction of http service

### PR DESCRIPTION
**Pull request by d3r1w**
# HTTP Service

The http service aims to simplify and secure the handling with the http request and response. Furthermore it enables developers to test the cookie handling, which was previously close to impossible.

## Features
* PSR-7 compliant request and response handling.
* Cookie handling

### guzzlehttp/psr7
The currently used PSR-7 implementation used by the http service.

### ralouphie/getallheaders
A polly fill for the getallheader function which is used by the guzzlehttp/psr7. This pollyfill is needed if no Apache is present in the system setup.

### psr/http-message
The PSR-7 interfaces designed by the PHP-FIG.

### dflydev/fig-cookies
The current implementation to handle the cookies. This library was designed to work with the http server request interface of the PSR-7.

### mockery/mockery
Mockery provides a convenient api to create mock objects for the unit tests.

### CODE REVIEW

File: src/HTTP/Cookies/CookieWrapper.php
Comment (shecken) [inline src/HTTP/Cookies/CookieWrapper.php]: In my opinion this isn't an immutable implementation of Cookie. In _with_ functions the property _cookie_ is changed and then returned new object. So we have a changed source object and a new instance and both are equal.

File: src/HTTP/Cookies/CookieJarWrapper.php
Comment (shecken) [inline src/HTTP/Cookies/CookieJarWrapper.php]: Same problem as in CookieWrapper. In this case _with_ and _without_ never create a new instance of CookieJarWrapper. We only get a changed source object.

**Comment by d3r1w** [comment]
Hi @shecken 
You are absolutely right I will fix the the problem as soon as possible.

**Comment by d3r1w** [comment]
The cookie wrapper and cookie jar wrapper should now be immutable as expected. I hope the changes are to your liking.

**Comment by chfsx** [comment]
this service is needed to make everything concerning requests and cookies testable using unitttest. this is needed to test the WebAccessChecker and FileDelivery more reliable that we can currently.

Have a look at the Service-README: https://github.com/studer-raimann/ILIAS/blob/feature/5-3/IntroductionOfHTTPService/src/HTTP/README.md

I kindly ask all core-developers to comment this PR if needed. After two weeks I will propose to merge this request during JourFixe if there are at least two core-developers supporting the request and if there are voices against it, if more than 2/3 core developers support it. thanks in advance!

File: src/HTTP/Cookies/CookieWrapper.php
Comment (Amstutz) [reply src/HTTP/Cookies/CookieWrapper.php]: @shecken: I am not sure about that. Let's look closely at an example:
[Edit]: Sry, I just saw that this has been changed in a previous commit already. I commented a later version.
```
public function withValue($value = null)
{
	return new self($this->cookie->withValue($value));
}
```
$this->cookie is an instance of setCookie and setCookie has the following withValue implementation (see: [SetCookie.php](https://github.com/dflydev/dflydev-fig-cookies/blob/master/src/Dflydev/FigCookies/SetCookie.php)):

```
    public function withValue($value = null)
    {
        $clone = clone($this);
        $clone->value = $value;
        return $clone;
    }
```
so $this->cookie->withValue creates a completely new setCookie and directly passed to the constructor of the new CookieWrapper without changing the source object.

However, I do not see why the constructor is used here. Would it not be more efficient to perform (if I remember correctly, constructors are more expensive than clone):

```
public function withValue($value = null)
{
        $clone = clone($this);
        $clone->cookie = $this->cookie->withValue($value);
        return $clone;
}
```

File: src/HTTP/Cookies/CookieJarWrapper.php
Comment (Amstutz) [reply src/HTTP/Cookies/CookieJarWrapper.php]: @shecken: I do not agree. Same reasoning as bellow in CookieWrapper. 
[Edit]: Sry, I just saw that this has been changed in a previous commit already. I commented a later version.

However, why are constructors instead of clone used?

File: src/HTTP/Factory.php
Comment (Amstutz) [inline src/HTTP/Factory.php]: Maybe a tremendously stupid question, but while browsing this I did not really get that: Where and how will this factory be implemented? By whom?

File: src/HTTP/Factory.php
Comment (Amstutz) [inline src/HTTP/Factory.php]: A function returning void seems strange on an interface named "Factory". Consider renaming the interface.

File: src/HTTP/Factory.php
Comment (Amstutz) [inline src/HTTP/Factory.php]: A function returning void seems strange on an interface named "Factory". Consider renaming the interface.

File: src/HTTP/Factory.php
Comment (Amstutz) [inline src/HTTP/Factory.php]: A function returning void seems strange on an interface named "Factory". Consider renaming the interface.

**Review by Amstutz — APPROVED** [review]
Thank you very much for this PR. I think the overall quality and documentation is on a very high level. Thanks also for providing some unit tests a long with it. 

In general I approve this PR. There are some suggestions for improvement, but they are of minor priority to me. 

Consider adding some test sets that potential implementation of the http factory have to pass.

File: src/HTTP/Cookies/CookieWrapper.php
Comment (shecken) [reply src/HTTP/Cookies/CookieWrapper.php]: @Amstutz 
I made this comment about 11 days ago. Since that d3r1w changed the implementation in immutable. So my comments are obsolete.

You're right. Looks like _clone_ is a little cheaper then _new_. I don't know how often it will be called in this case, but i think it could be negligible. Benchmark runs i found shows differences about 0,2 seconds if cloning 4 object in 100.000 cycles.

File: src/HTTP/Factory.php
Comment (d3r1w) [reply src/HTTP/Factory.php]: Hi @Amstutz 
I am sorry, I forgot to push the implementation of the factory. I will do that as fast as possible.

**clone vs constructor**
I was not aware of a performance benefit of the clone method. Is there a "considerable" performance difference between the two methods? If thats the case, I will change the implementation.

File: src/HTTP/Factory.php
Comment (d3r1w) [reply src/HTTP/Factory.php]: Yes you are absolutely right, how would you name the interface? Maybe just HTTPServiceInterface?

Comment (klees) [reply src/HTTP/Factory.php]: Hi @d3r1w,

this is the binding to be able to support the current handling of HTTP in ILIAS we once talked about, right?

From the perspective of the proposed FIG-Interfaces for HTTP services, request handling (i.e. our app) should have the form of a function from Request to Response. This makes it possible to e.g. use some 3rd party middleware somewhere in the chain that performs the request handling. The kind of object you are proposing with this `Factory` somehow circumvents this style of request handling, as it introduces the possibility to store some request at some time or get another request somewhere else. This destroys some of the predictability that is implied by the Request -> Response model.
I see, however, how we require such a facility to be able to use this HTTP-service at the moment, i.e. make use of it in a world where the request is stored in some super globals and responses are made via the global template or echo. Maybe we should name this facility `GlobalHTTPState` to have a reminder for this.

Thx for the good work, though.

Regards!

File: src/HTTP/Cookies/Cookie.php
Comment (klees) [inline src/HTTP/Cookies/Cookie.php]: `rememberForLongTime`? 5 years is not forever, I can tell you...

File: src/HTTP/Cookies/Cookie.php
Comment (klees) [inline src/HTTP/Cookies/Cookie.php]: What does it mean if I pass `null` here? Is this `expires`? Why have both then?

File: src/HTTP/Cookies/Cookie.php
Comment (klees) [inline src/HTTP/Cookies/Cookie.php]: We should not have static functions on interfaces, because no one could call them. Move these `create*` methods to a factory for cookies and you are fine.

File: src/HTTP/Cookies/Cookie.php
Comment (klees) [inline src/HTTP/Cookies/Cookie.php:180]: What is the string representation of the cookie? "Key: value"?

File: src/HTTP/Cookies/CookieJar.php
Comment (klees) [inline src/HTTP/Cookies/CookieJar.php:8]: What is a cookie jar? Is this the place the cookies are stored at? This automatically created docu is less then helpful. One can indeed see, that this is an interface named CookieJar. It would be nice to know what it is, or models, or represents, ...

File: src/HTTP/Cookies/CookieJar.php
Comment (klees) [inline src/HTTP/Cookies/CookieJar.php]: `is the current jar.` -> `in the jar.`

File: src/HTTP/Factory.php
Comment (Amstutz) [reply src/HTTP/Factory.php]: +1 for `GlobalHTTPState`

File: src/HTTP/Cookies/CookieJar.php
Comment (klees) [inline src/HTTP/Cookies/CookieJar.php]: See static functions in the cookie interface. Move statics to a factory.

File: src/HTTP/Request/RequestFactory.php
Comment (klees) [inline src/HTTP/Request/RequestFactory.php]: This makes it harder to test objects using this. Please remove the `static`.

File: src/HTTP/Response/ResponseFactory.php
Comment (klees) [inline src/HTTP/Response/ResponseFactory.php]: This makes it harder to test objects using this. Please remove the static.

File: src/HTTP/Factory.php
Comment (Amstutz) [reply src/HTTP/Factory.php]: @d3r1w Concerning the performance of clone vs constructor. The gain seems not to be huge. @shecken wrote above:

> You're right. Looks like clone is a little cheaper then new. I don't know how often it will be called in this case, but i think it could be negligible. Benchmark runs i found shows differences about 0,2 seconds if cloning 4 object in 100.000 cycles.

I would personally prefer clone. But it seems to be more about personal preference in this case than performance. However, if you write code that will clone a lot (such as in the UI-Framework), definitely use clone instead of new.

However, we should not forget, that 400'000 sound like a lot, but might be quickly reached in an application like ILIAS AND 0.2s is definitely a huge performance loss.

File: src/HTTP/Response/rendering/DefaultResponseRenderingStrategy.php
Comment (klees) [inline src/HTTP/Response/rendering/DefaultResponseRenderingStrategy.php]: Why do you use a lowercase namespace-name?

File: src/HTTP/Response/rendering/ResponseRenderingStrategy.php
Comment (klees) [inline src/HTTP/Response/rendering/ResponseRenderingStrategy.php]: The name of this interface is wrong. As the comment for `renderResponse` and the implementation for `DefaultResponseRenderingStrategy` show, this is a `ResponseSenderStrategy`. It certainly does more then rendering.

Why is this a `Strategy` btw? How is this different from a `ResponseSender`?

**Review by klees — CHANGES_REQUESTED** [review]
LGTM in general, thx for the good work. My comments contain some nitpicking (as usual), which should be easy to fix.

File: src/HTTP/Factory.php
Comment (d3r1w) [reply src/HTTP/Factory.php]: Hi @klees
Thanks again for your feedback. Yes that's the one we talked about.

I can only agree with your opinion. Should I rename the interface to GlobalHttpStateInterface and the implementation of the interface to GlobalHttpState or just the interface to GlobalHttpState? (The current name of the implementation is "HTTPServices".)

Comment (klees) [reply src/HTTP/Factory.php]: I prefer dropping the `Interface` from the interface names and looking for another name for the implementation. The interface is what you would be type hinting on, so we save some chars there.

File: src/HTTP/Cookies/Cookie.php
Comment (d3r1w) [reply src/HTTP/Cookies/Cookie.php]: A cookie can have a date of expiry or a max-age or both which is a bit confusing but possible. The null parameter is used to erase the date of expiry out of the cookie. The expires method is used set the date of expiry to a date in past. I will update the doc comment to clarify the usage of the method.

File: src/HTTP/Cookies/Cookie.php
Comment (d3r1w) [reply src/HTTP/Cookies/Cookie.php]: I never thought that a cookie could survive over 5 years within a browser. To address that, I will rename the method as you suggested and update the doc comment to clarify the behaviour.

File: src/HTTP/Factory.php
Comment (d3r1w) [reply src/HTTP/Factory.php]: @Amstutz 
Sorry, I somehow missed the answer of @shecken. That is a huge difference, I will change the implementation as soon as possible. Thanks @shecken for the test!

File: src/HTTP/Response/rendering/ResponseRenderingStrategy.php
Comment (d3r1w) [reply src/HTTP/Response/rendering/ResponseRenderingStrategy.php]: @klees 
That is a good point, I will update the names within the package.

They are strategies because of the strategy pattern used within the service. The senders are representing the strategies, the sender interface represent the abstraction and the HTTPServices class represents the context. I adopted the naming scheme of the factories.

File: src/DI/Container.php
Comment (klees) [inline src/DI/Container.php]: Hi @d3r1w,

this defeats the purpose of the dependency injection container. No dependency injected here, static binding to the two implementations instead. Do you understand the purpose and mechanism of DI?

The `UIServices` from [l.109](https://github.com/ILIAS-eLearning/ILIAS/pull/466/files#diff-f46de88208848386905f0e063e513b62L109) is just for structuring the access to the DIC-properties further and does not contain the real UI-services on its own. These are added to the container in ilInitialisation, which would imo would be the proper place to add the HTTPServices and the two required implementations as well.

I'm not convinced we should expose the GlobalHTTPState via a method on the Container. Why not inject it in the services that requires the state as well?

Best regards!

File: src/DI/HTTPServices.php
Comment (klees) [inline src/DI/HTTPServices.php]: This totally defeats the purpose of the DIC. It is there to define dependencies, not to store application state.

**Review by klees — CHANGES_REQUESTED** [review]
There seem to be some missunderstandings regarding the DIC. If you need to discuss that stuff I'm available for a Skype session.

File: src/DI/Container.php
Comment (d3r1w) [reply src/DI/Container.php]: Hi @klees 

Yes or at least I hope so. You are absolutely right.
That is the first time I used a DIC which can't resolve the dependencies by it self, that confused me a bit.
Is that the proper way of doing that? (ilInitialisation)

```
        $container['http.request_factory'] = function ($c) {
            return new \ILIAS\HTTP\Request\RequestFactoryImpl();
        };

        $container['http.response_factory'] = function ($c) {
            return new \ILIAS\HTTP\Response\ResponseFactoryImpl();
        };

        $container['http.cookie_jar_factory'] = function ($c) {
            return new \ILIAS\HTTP\Cookies\CookieJarFactoryImpl();
        };

        $container['http.response_sender_strategy'] = function ($c) {
            return new \ILIAS\HTTP\Response\Sender\DefaultResponseSenderStrategy();
        };

        $container['http'] = function ($c) {
            return new \ILIAS\DI\HTTPServices(
                $c['http.response_sender_strategy'],
                $c['http.cookie_jar_factory'],
                $c['http.request_factory'],
                $c['http.response_factory']
            );
        };
```
I also moved the request and response out of the DIC into the GlobalHttpState.
The new global HTTPServices constructor would now look like this. (without the wrong indentation)

```
public function __construct(
        ResponseSenderStrategy $senderStrategy,
        CookieJarFactory $cookieJarFactory,
        RequestFactory $requestFactory,
        ResponseFactory $responseFactory
    )
	{
		$this->sender = $senderStrategy;
		$this->$cookieJarFactory = $cookieJarFactory;

		$this->requestFactory = $responseFactory;
		$this->responseFactory = $responseFactory;
	}
```
About the injection of the GlobalHttpState.
That is a nice idea, but does that work for all services? The Web Access Checker for example has an own entry point and is never initialized by ILIAS.

I will implement the changes if they are ok. Otherwise, I would be glad to discuss further actions via Skype.

Comment (klees) [reply src/DI/Container.php]: LGTM.

I think the own entry point for the WAC makes it even easier to inject the GlobalHTTPState, but I might be wrong. You could always use $DIC["http"].

The thing that lets me argue about the exposal of the GlobalHTTPState via a method on the DIC is twofold. The exposal of said methods kind of promotes a dependency antipattern (service locator). We still need that methods to make a migration easier. That is why we decided to only expose some services that are used a lot to allow your IDEs to display hints. Second thing is, that the general existence of the GlobalHTTPState is kind of an antipattern to the request -> response model implied by PSR 7. Both points aren't too crucial, though, so this is more a "but..." then a "NO!".

Comment (d3r1w) [reply src/DI/Container.php]: Thanks for the quick feedback.

I hope you are right, it would simplify my next task(#467).

That are really good arguments and I fully agree with your opinion. Therefore, I will remove the  method from the Container class to (hopefully) reduce the direct usage of the GlobalHttpState.

File: src/DI/Container.php
Comment (klees) [inline src/DI/Container.php]: Remove these please.

File: src/DI/HTTPServices.php
Comment (klees) [inline src/DI/HTTPServices.php]: Please use tabs for indentation. I'm would prefer 4 spaces as well, but [coding guidelines](http://www.ilias.de/docu/ilias.php?ref_id=42&from_page=29964&obj_id=202&cmd=layout&cmdClass=illmpresentationgui&cmdNode=fr&baseClass=ilLMPresentationGUI)...

You have spaces in some other files as well.

**Review by klees — APPROVED** [review]
Thx Nick, very nice, very brave. One of the biggest PRs so far, I think.

**Comment by d3r1w** [comment]
Thx for all your feedback!

@chfsx will push the composer dependencies after the PR is merged.

**Review by chfsx — APPROVED** [review]

**Comment by matthiaskunkel** [comment]
Jour Fixe, April 24, 2017: Highly appreciated and approved. Will be merged by Fabian Schmid. First maintainer of this service will be Fabian.

**Pull request merged**
